qemu/tests/single_driver_install.py

```diff
@@ -17,153 +40,96 @@ def run(test, params, env):
     try:
-        error_context.context("Show file extentions", logging.info)
-        show_file_extentions(session)
-        session.close()
-        session = reboot(vm)
-
-        uninstall_flag = params.get("need_uninstall", "no")
-        if uninstall_flag == "yes":
-            operation = "uninstall_driver"
-            error_context.context("Uninstall driver", logging.info)
-            install_driver(session, operation)
-            session = reboot(vm)
-
-        operation = "install_driver"
-        error_context.context("Install driver", logging.info)
-        install_driver(session, operation)
-        session = reboot(vm)
-
-        if uninstall_flag == "no":
-            operation = "verify_driver"
-            error_context.context("Verify driver is same as expected", logging.info)
-            install_driver(session, operation)
-
-    finally:
-        if session:
-            error_context.context("Get driver installation log", logging.info)
-            try:
-                get_installation_logs(session)
-            except (ShellCmdError, ShellProcessTerminatedError):
-                pass
-            session.close()
+        get_drive_letter = getattr(virtio_win, "drive_letter_%s" % media_type)
+        get_product_dirname = getattr(virtio_win,
+                                      "product_dirname_%s" % media_type)
+        get_arch_dirname = getattr(virtio_win, "arch_dirname_%s" % media_type)
+    except AttributeError:
+        test.error("Not supported virtio win media type '%s'", media_type)
+    viowin_ltr = get_drive_letter(session)
+    if not viowin_ltr:
+        test.error("Could not find virtio-win drive in guest")
+    guest_name = get_product_dirname(session)
+    if not guest_name:
+        test.error("Could not get product dirname of the vm")
+    guest_arch = get_arch_dirname(session)
+    if not guest_arch:
+        test.error("Could not get architecture dirname of the vm")
+
+    inf_middle_path = ("{name}\\{arch}" if media_type == "iso"
+                       else "{arch}\\{name}").format(name=guest_name,
+                                                     arch=guest_arch)
+    inf_find_cmd = 'dir /b /s %s\\%s.inf | findstr "\\%s\\\\"'
+    inf_find_cmd %= (viowin_ltr, driver_name, inf_middle_path)
+    inf_path = session.cmd(inf_find_cmd).strip()
+    logging.info("Found inf file '%s'", inf_path)
+
+    expected_ver = session.cmd("findstr DriverVer= %s" % inf_path)
+    expected_ver = expected_ver.strip().split(",", 1)[-1]
+    if not expected_ver:
+        test.error("Failed to find driver version from inf file")
+    logging.info("Target version is '%s'", expected_ver)
+
+    if params.get("need_uninstall", "no") == "yes":
+        error_context.context("Uninstalling previous installed driver",
+                              logging.info)
+        for inf_name in _pnpdrv_info(session, device_name, ["InfName"]):
+            uninst_store_cmd = "pnputil /f /d %s" % inf_name
```

Comment:
does it need judge if inf_name is None here as parse_list() may return []?

Reply:
It won't go into the loop if `_pnpdrv_info()` returns `[]`.